Report dependencies externalized with Dependency Extraction Plugin

tomalec opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
This PR adds a new option for plugin config `externalizedReportFile`, to report all dependencies exported for the current project.
I need it as I try to build a tool that helps to inspect, analyze and manage the versions being used for local unit tests, and that runs with WordPress.
To compare the versions used locally, with the versions used in the range of supported WP builds.


----
edit:
 This PR tries to address problem 5. from https://github.com/WordPress/gutenberg/issues/35630
> As a plugin developer, I don't even know what dependencies are extracted, as the list is maintained in the package repo, and not reported while bundling.

Thanks to this PR a developer would at least know which of their local dependencies are extracted and which are not. So for which packages do they need to anticipate different versions.

----


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
I run this code against 
https://github.com/woocommerce/google-listings-and-ads repo's webpack config with and without the new option.

0. ```
   git clone https://github.com/woocommerce/google-listings-and-ads.git
   cd google-listings-and-ads
   npm install
   ```
1. Replace `node_modules/@wordpress/dependency-extraction-webpack-plugin` with the changed one from this PR `packages/dependency-extraction-webpack-plugin`.

### Without the new option

2. Use the repo's Webpack config as is https://github.com/woocommerce/google-listings-and-ads/blob/4e69b169953096774dd0d957ba7a4fcecbbf2a89/webpack.config.js#L79-L83
   ```js
		new DependencyExtractionWebpackPlugin( {
			injectPolyfill: true,
			requestToExternal,
			requestToHandle,
		} ),
   ```
3. `npm run build`
4. Check that the `/js/build/index.js` bundle is created as before.

### With the new option

2. Add the `reportExternalized` option to the `/webpack.config.js#L79-L83`:
   ```js
		new DependencyExtractionWebpackPlugin( {
			injectPolyfill: true,
			requestToExternal,
			requestToHandle,
			externalizedReportFile: 'externalized.json',
		} ),
   ```
3. `npm run build`
4. Check that the `/js/build/index.js`  bundle is created as before.
5. Check the `/js/build/externalized.js` is created and contains:
	```json
	["@wordpress/i18n","@wordpress/hooks","@woocommerce/navigation","@woocommerce/wc-admin-settings","@wordpress/element","@woocommerce/components","@wordpress/api-fetch","@wordpress/data","@wordpress/deprecated","@babel/runtime/regenerator","@wordpress/date","lodash","react","@wordpress/html-entities","moment","@wordpress/is-shallow-equal","@wordpress/keycodes","@wordpress/a11y","@wordpress/priority-queue","@wordpress/dom","react-dom","@wordpress/warning","@wordpress/rich-text","@wordpress/data-controls","@wordpress/url"]
	```


## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested. - tested manually, I was not able to run tests for this package only, and `npm install` for entire repo fails on my machine
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [n/a] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [n/a] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

tomalec committed:
Add `externalizedReportFile` option to DEWP,

to report all dependencies extracted for the burrent build.

tomalec committed:
Add CHANGELOG.md entry.

tomalec committed:
Fix typo.

gziolo commented:
Thank you for opening this Pull Request.

> I need it as I try to build a tool that helps to inspect, analyze and manage the versions being used for local unit tests, and that runs with WordPress  (See articles: pb0Spc-1LK-p2, p9oQ9f-zJ-p2).

How people can access those articles? It's hard to understand how the changes proposed would benefit the WordPress project and the community without this context.

The build process in Gutenberg and WordPress core replaces all WP dependencies with `wp.*` globals. In practice, the versions of WP packages or core libraries like `lodash` or `react` used in the development will always differ from those in WordPress because they depend on the version of WordPress core and whether a Gutenberg plugin is installed.

tomalec commented:
> How people can access those articles? It's hard to understand how the changes proposed would benefit the WordPress project and the community without this context.

I created an issue in this repo https://github.com/WordPress/gutenberg/issues/35630
I hope it gives enough context and rationale to investigate this problem, and partial solution.

This PR tries to address [problem 5. stated there](https://github.com/WordPress/gutenberg/issues/35630#issue-1022052698)
> As a plugin developer, I don't even know **what dependencies are extracted**, as the list is maintained in the package repo, and not reported while bundling.
That may result in unexpected behavior and error. Checking that again requires even more manual effort. As it requires digging into the source code of DEWP at a given version and manually comparing packages list.

----

> The build process in Gutenberg and WordPress core replaces all WP dependencies with `wp.*` globals. In practice, the versions of WP packages or core libraries like `lodash` or `react` used in the development will always differ from those in WordPress because they depend on the version of WordPress core and whether a Gutenberg plugin is installed.

That's exactly the DevX problem I'm trying to address here.


---

Thanks to this PR a developer would at least know which of their local dependencies are extracted and which are not. So for which packages do they need to anticipate different versions.

jsnajdr commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -23,6 +23,7 @@ class DependencyExtractionWebpackPlugin {
 			{
 				combineAssets: false,
 				combinedOutputFile: null,
+				externalizedReportFile: undefined,
```
Just to be consistent with the `combinedOutputFile` option, the default value should be `null` and the type should be `combinedOutputFile?: string | null`. Or, as defaulting the option to `undefined` doesn't really have a visible impact, omit it completely?

gziolo replied:
Should it align with `combineAssets` (enables feature) and `combinedOutputFile` (provides a custom file name) pair?

tomalec replied:
> Should it align with `combineAssets` (enables feature) and `combinedOutputFile` (provides a custom file name) pair?

I get the point of consistency and agree the new API should be consistent with the old one.

I'm just not sure I understand the value of having those. Personally, as a consumer, I like the API to be as small as possible, while delivering the same value. As that usually reduces the number of new things I need to learn, bloat in my config/code, the confusion regarding a number of different combinations, and a chance to make a mistake.

I that case I'd wonder what is the expected outcome following settings?
- `combineAssets/reportExternalized: false | undefined` with `combinedOutputFile/reportExternalizedFile: 'foo.bar'` 
- `combineAssets/reportExternalized: true` with `combinedOutputFile/reportExternalizedFile` not set, `null`, and `undefined`

jsnajdr reviewed (approved):
Works well for me :+1:

gziolo commented on packages/dependency-extraction-webpack-plugin/lib/types.d.ts:
```diff
@@ -15,4 +15,5 @@ declare interface DependencyExtractionWebpackPluginOptions {
 	requestToHandle?: ( request: string ) => string | undefined;
 	combinedOutputFile?: string | null;
 	combineAssets?: boolean;
+    externalizedReportFile?: string | undefined;
```
@jsnajdr, is Prettier ignoring `*.d.ts` files?

jsnajdr replied:
It should format them. When I ran `wp-scripts format` in my CLI now, it got formatted:
```
$ ./node_modules/.bin/wp-scripts format packages/dependency-extraction-webpack-plugin
packages/dependency-extraction-webpack-plugin/lib/index.js 94ms
packages/dependency-extraction-webpack-plugin/lib/types.d.ts 143ms
packages/dependency-extraction-webpack-plugin/lib/util.js 14ms
```

gziolo replied:
I think we skip `.d.ts` from linting then. I now remember we had some issues with that in the past. 😄

gziolo committed:
Update types.d.ts

gziolo commented:
> I run this code against
https://github.com/woocommerce/google-listings-and-ads repo's webpack config with and without the new option.

What do I have to provide to enable this option? Can you share a full example?

jsnajdr commented:
> What do I have to provide to enable this option? Can you share a full example?

I tested this new option by modifying the `tools/webpack/shared.js` config file and running `npm run build`. A `report.json` file was created.

gziolo commented:
Thinking a bit more about it, if you want to see a full list of externalized dependencies you can enable `combineAssets` flag and it will report all of them in a single file grouped by chunks. It is also what WordPress core uses:

https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/assets/script-loader-packages.php

tomalec commented:
> What do I have to provide to enable this option? Can you share a full example?

I added more elaborate steps to reproduce in the OP https://github.com/WordPress/gutenberg/pull/35106#issue-742060241

Is that clearer now?

tomalec commented:
> Thinking a bit more about it, if you want to see a full list of externalized dependencies you can enable `combineAssets` flag and it will report all of them in a single file grouped by chunks. It is also what WordPress core uses:
>
> WordPress/wordpress-develop@master/src/wp-includes/assets/script-loader-packages.php

Interesting. I didn't think `combineAssets` flag could serve to report dependency.
After reading the API docs:
> By default, one asset file is created for each entry point. When this flag is set to true, all information about assets is combined into a single assets.(json|php) file generated in the output directory.

I thought it is to change the behavior of generating assets separately for each entry point to generating a single file. So to change the way it actually exposes the assets to be used on runtime.

Also, for
```js
new DependencyExtractionWebpackPlugin( {
	  combineAssets: false,
	  outputFormat: 'json',
} );
```
it generates this list of internal identifiers
```json
{
	"index.js": {
		"dependencies": [
			"lodash",
			"moment",
			"react",
			"react-dom",
			"wc-components",
			"wc-navigation",
			"wc-settings",
			"wp-a11y",
			"wp-api-fetch",
			"wp-data",
			"wp-data-controls",
			"wp-date",
			"wp-deprecated",
			"wp-dom",
			"wp-element",
			"wp-hooks",
			"wp-html-entities",
			"wp-i18n",
			"wp-is-shallow-equal",
			"wp-keycodes",
			"wp-polyfill",
			"wp-priority-queue",
			"wp-rich-text",
			"wp-url",
			"wp-warning"
		],
	// …
```
instead of ~npm package names~ module specifiers used in the local (plugin's) codebase
```json
[
	"@wordpress/i18n",
	"@wordpress/hooks",
	"@woocommerce/wc-admin-settings",
	"@woocommerce/navigation",
	"@wordpress/element",
	"@woocommerce/components",
	"@wordpress/data-controls",
	"@wordpress/data",
	"@wordpress/api-fetch",
	"@babel/runtime/regenerator",
	"@wordpress/url",
	"@wordpress/date",
	"lodash",
	"moment",
	"@wordpress/deprecated",
	"@wordpress/a11y",
	"@wordpress/html-entities",
	"react",
	"@wordpress/keycodes",
	"@wordpress/priority-queue",
	"react-dom",
	"@wordpress/dom",
	"@wordpress/is-shallow-equal",
	"@wordpress/rich-text",
	"@wordpress/warning"
]
```

But that's pretty close, do you think we can somewhat merge those two together? 
Like serializing the actual `Map< externalizedDependency, scriptDependency >`?

jsnajdr commented:
> if you want to see a full list of externalized dependencies you can enable `combineAssets` flag

Seems to me that the purposes of `combineAssets` and `externalizedReportFile` are quite different and it's not advisable to mix them:
- `combineAssets` specifies how that `*.asset.php` files should look like, and these files are part of the published artifact. WordPress PHP code that loads the scripts depends on them.
- `externalizedReportFile` is not supposed to be published -- it's an intermediate report/log/stats file that's supposed to be an input to some next step in the build pipeline. Or helps the developer with debugging their dependencies.

tomalec commented:
@gziolo What's your take on the above comments?

gziolo commented on packages/dependency-extraction-webpack-plugin/README.md:
```diff
@@ -135,6 +135,13 @@ Pass `useDefaults: false` to disable the default request handling.
 
+##### `externalizedReportFile`
+
+-   Type: string
```
How about it’s both `string` and `boolean`? This way it’s `false` by default, it’s using some predefined path when `true` and a custom path otherwise.

tomalec replied:
@gziolo I implemented it in https://github.com/WordPress/gutenberg/pull/35106/commits/81b2d9b185a14105049d27b25ea1e6ba3bf70876
and change names a little bit. Could you confirm the implementation and default name is fine to you?

Option name: `externalizedReport`
Option type: `boolean | string`
Default value: `false`
Default file name: `'externalized-dependencies.json'`

gziolo reviewed (approved):
I think we should focus on bringing ES Modules and import maps to WordPress core to eventually make this plugin obsolete. It’s going to be a long process so if you think it’s helpful to add this feature then let’s land it.

tomalec committed:
Make DEWP's `externalizedReport` option accept booleans.

Use the default filename - `'externalized-dependencies.json'` for `true`.
Addresses https://github.com/WordPress/gutenberg/pull/35106#pullrequestreview-809564706

gziolo commented:
One of the Continues Integration jobs fails even on subsequent re-tries: https://github.com/WordPress/gutenberg/runs/4261297083?check_suite_focus=true

Can you check if rebasing with `trunk` resolves it? I doubt there's anything wrong with this branch.

tomalec commented:
> Can you check if rebasing with `trunk` resolves it? I doubt there's anything wrong with this branch.


Thanks, @gziolo . Rebased, everything is green now.

tomalec commented:
@gziolo is there any date when this (and [two other PRs](https://github.com/WordPress/gutenberg/commits/30aab8a2ae16045820d306b2a9380d194c1d4396/packages/dependency-extraction-webpack-plugin)) could be released to npm? - so it could be easily used outside

gziolo commented:
See: https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#synchronizing-wordpress-trunk

> For each Gutenberg plugin release, WordPress trunk should be synchronized. Note that the WordPress trunk branch can be closed or in "feature-freeze" mode. Usually, this happens between the first beta and the first RC of the WordPress release cycle. During this period, the Gutenberg plugin releases should not be synchronized with WordPress Core.

We are in the process of releasing the next major WordPress version (5.9) so we don't do regular npm publishing every two weeks.

I plan to run [the development release](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#development-releases) with the `next` dist-tag later this week in case someone wants to use the most recent version earlier.

tomalec commented:
Thank you for the info!

BTW, I made a small improvement for this feature in https://github.com/WordPress/gutenberg/pull/37377, maybe it could get into that release as well?

gziolo commented:
It's now available for testing on npm with:
```
npm install --save-dev @wordpress/dependency-extraction-webpack-plugin@next
```

The stable version should be available in the second half of January.

tomalec commented:
Awesome, thanks for the heads-up :)
